src/Tolerance/Operation/Runner/RetryPromiseOperationRunner.php

```diff
@@ -0,0 +1,108 @@
+<?php
+
+/*
+ * This file is part of the Tolerance package.
+ *
+ * (c) Samuel ROZE <samuel.roze@gmail.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Tolerance\Operation\Runner;
+
+use Tolerance\Operation\Exception\UnsupportedOperation;
+use Tolerance\Operation\Operation;
+use Tolerance\Operation\PromiseOperation;
+use Tolerance\Operation\RetryEvaluator\AlwaysRetryEvaluator;
+use Tolerance\Operation\RetryEvaluator\NeverRetryEvaluator;
+use Tolerance\Operation\RetryEvaluator\RetryEvaluator;
+use Tolerance\Waiter\StatefulWaiter;
+use Tolerance\Waiter\Waiter;
+use Tolerance\Waiter\WaiterException;
+
+class RetryPromiseOperationRunner implements OperationRunner
+{
+    /**
+     * @var Waiter
+     */
+    private $waitStrategy;
+    /**
+     * @var RetryEvaluator
+     */
+    private $fulfilledEvaluator;
+    /**
+     * @var RetryEvaluator
+     */
+    private $rejectedEvaluator;
+
+    /**
+     * @param Waiter $waitStrategy
+     * @param RetryEvaluator|null $fulfilledEvaluator
+     * @param RetryEvaluator|null $rejectedEvaluator
+     */
+    public function __construct(Waiter $waitStrategy, RetryEvaluator $fulfilledEvaluator = null, RetryEvaluator $rejectedEvaluator = null)
+    {
+        $this->waitStrategy = $waitStrategy;
+        $this->fulfilledEvaluator = $fulfilledEvaluator ?: new NeverRetryEvaluator();
```

Comment:
Yes. With Guzzle, the promise is fulfilled when a response is available (even 4xx or 5xx):

```
and returns a GuzzleHttp\Promise\PromiseInterface that is fulfilled with a Psr\Http\Message\ResponseInterface or rejected with an exception.
```
(from http://docs.guzzlephp.org/en/latest/handlers-and-middleware.html#creating-a-handler)

That's why I added a Guzzle promise RetryEvaluator: https://github.com/iamluc/Tolerance/blob/promise/src/Tolerance/Bridge/Guzzle/Operation/RetryEvaluator/PromiseRetryEvaluator.php